aclk/https_client.c

```diff
@@ -129,118 +147,408 @@ static int parse_http_response(rbuf_t buf, http_parse_ctx *parse_ctx)
+    struct timeval timeout = { .tv_sec = request->timeout_s, .tv_usec = 0 };
+
+    https_req_ctx_t *ctx = callocz(1, sizeof(https_req_ctx_t));
+    ctx->req_start_time = now_realtime_sec();
+
+    ctx->buf_rx = rbuf_create(RX_BUFFER_SIZE);
+    if (!ctx->buf_rx) {
+        error("Couldn't allocate buffer for RX data");
+        rc = 20;
+        goto exit_req_ctx;
     }
 
-    ret = SSL_write(ssl, b, len);
-    if (ret <= 0)
-    {
-        error("SSL_write() failed with err=%d", ret);
-        goto exit_SSL;
+    snprintf(connect_port_str, PORT_STR_MAX_BYTES, "%d", connect_port);
+
+    ctx->sock = connect_to_this_ip46(IPPROTO_TCP, SOCK_STREAM, connect_host, 0, connect_port_str, &timeout);
+    if (ctx->sock < 0) {
+        error("Error connecting TCP socket to \"%s\"", connect_host);
+        rc = 30;
+        goto exit_buf_rx;
     }
 
-    b[0] = 0;
+    if (fcntl(ctx->sock, F_SETFL, fcntl(ctx->sock, F_GETFL, 0) | O_NONBLOCK) == -1) {
+        error("Error setting O_NONBLOCK to TCP socket.");
+        rc = 40;
+        goto exit_sock;
+    }
 
-    do {
-        ptr = rbuf_get_linear_insert_range(buffer, &len);
-        ret = SSL_read(ssl, ptr, len - 1);
-        if (ret)
-            rbuf_bump_head(buffer, ret);
-        if (ret <= 0)
-        {
-            error("No response available - SSL_read()=%d", ret);
-            goto exit_FULL;
+    ctx->poll_fd.fd = ctx->sock;
+
+    if (request->proxy_host) {
+        https_req_t req = HTTPS_REQ_T_INITIALIZER;
+        req.request_type = HTTP_REQ_CONNECT;
+        req.timeout_s = request->timeout_s;
+        req.host = request->host;
+        req.port = request->port;
+        req.url = request->url;
+        ctx->request = &req;
+        if (handle_http_request(ctx)) {
+            error("Failed to CONNECT with proxy");
+            rc = 45;
+            goto exit_sock;
         }
-    } while (!(ret = parse_http_response(buffer, &parse_ctx)));
+        if (ctx->parse_ctx.http_code != 200) {
+            error("Proxy didn't return 200 OK (got %d)", ctx->parse_ctx.http_code);
+            rc = 46;
+            goto exit_sock;
+        }
+        info("Proxy accepted CONNECT upgrade");
+    }
+    ctx->request = request;
 
-    if (ret != PARSE_SUCCESS) {
-        error("Error parsing HTTP response");
-        goto exit_FULL;
+    ctx->ssl_ctx = security_initialize_openssl_client();
+    if (ctx->ssl_ctx==NULL) {
+        error("Cannot allocate SSL context");
+        rc = 50;
+        goto exit_sock;
     }
 
-    if (parse_ctx.http_code < 200 || parse_ctx.http_code >= 300) {
-        error("HTTP Response not Success (got %d)", parse_ctx.http_code);
-        goto exit_FULL;
+    ctx->ssl = SSL_new(ctx->ssl_ctx);
+    if (ctx->ssl==NULL) {
+        error("Cannot allocate SSL");
+        rc = 60;
+        goto exit_CTX;
+    }
+
+    SSL_set_fd(ctx->ssl, ctx->sock);
+    ret = SSL_connect(ctx->ssl);
+    if (ret != -1 && ret != 1) {
```

Comment:
Same check twice?

Reply:
it is -1 and 1

Reply:
to clarify one is instant connection success and another is failure sometimes expected if the underlying socket is nonblocking... in which case we check for specific error and continue